src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferStreamReadAsyncMemoryOverloadsTests.cs

```diff
@@ -0,0 +1,172 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Testing;
+using Xunit;
+
+namespace Microsoft.NetCore.Analyzers.Runtime.UnitTests
+{
+    public class PreferStreamReadAsyncMemoryOverloadsTest : PreferStreamAsyncMemoryOverloadsTestBase
+    {
+        #region No diagnostic - Analyzer
+
+        [Fact]
+        public async Task NoDiagnosticAnalyzer_Read()
+        {
+            await VerifyAnalyzerAsync50(@"
+using System;
+using System.IO;
+using System.Threading;
+class C
+{
+    void M()
+    {
+        using (FileStream fs = File.Open(""file.txt"", FileMode.Open))
+        {
+            byte[] result = new byte[fs.Length];
+            fs.Read(result, 0, (int)fs.Length);
+        }
+    }
+}
+            ");
+        }
+
+        [Fact]
+        public async Task NoDiagnosticAnalyzer_ReadAsyncWithAsMemory()
+        {
+            await VerifyAnalyzerAsync50(@"
+using System;
+using System.IO;
+using System.Threading;
+class C
+{
+    async void M()
+    {
+        using (FileStream fs = File.Open(""file.txt"", FileMode.Open))
+        {
+            byte[] result = new byte[fs.Length];
+            await fs.ReadAsync(result.AsMemory(), new CancellationToken());
+        }
+    }
+}
+            ");
+        }
+
+        #endregion
+
+        #region Diagnostic - Analyzer
+
+        [Fact]
+        public async Task DiagnosticAnalyzer_Basic()
+        {
+            await VerifyAnalyzerAsync50(@"
+using System;
+using System.IO;
+using System.Threading;
+class C
+{
+    async void M()
+    {
+        using (FileStream fs = File.Open(""file.txt"", FileMode.Open))
+        {
+            byte[] result = new byte[fs.Length];
+            await fs.ReadAsync(result, 0, (int)fs.Length);
+        }
+    }
+}
+            ", GetCSharpResult(12, 19));
```

Comment:
Please switch to unit tests which uses markup based annotations instead of magic line/column numbers. For example, here you can use `[|await fs.ReadAsync(result, 0, (int)fs.Length)|];` in the test code to validate that the test generates a single diagnostic with the annotated span. You don't need to pass in `GetCSharpResult` to `VerifyAnalyzerAsync50` invocation

Reply:
Let me correct myself - given you have multiple diagnostics being reported by the analyzer, you would need to prefix the annotation with the diagnostic ID: `{|CA18xx:await fs.ReadAsync(result, 0, (int)fs.Length)|}`. For an example, see https://github.com/dotnet/roslyn-analyzers/blob/6b16c9c25464f5f77d33da7a62f803857d4d3d6a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseAsSpanInsteadOfRangeIndexerTests.cs#L115

Reply:
Well, looking more closely, both the descriptors used by your analyzer have same diagnostic ID, but different messages. I don't know if it is possible to use markup based annotations here. @sharwell?

Reply:
@mavasani I tried using `[| ... |]` and removed the magic line/column numbers, and all the unit tests expecting a diagnostic started failing with this message:

```
Message:
System.InvalidOperationException:
Markup syntax can only omit the diagnostic ID if the first analyzer only supports a single diagnostic.
To customize the default value, override AnalyzerTest<TVerifier>.GetDefaultDiagnostic or specify MarkupOptions.UseFirstDescriptor.
```

Reply:
@mavasani @sharwell I tried your suggestion of enclosing the code sections with a diagnostic with `{|CA1834:code|}` and all those tests failed with a message like this:

```
Message: 
System.InvalidOperationException:
Multiple diagnostic descriptors with ID CA1834 were found.
Use the explicitly diagnostic creation syntax or specify MarkupOptions.UseFirstDescriptor to use the first matching diagnostic.

  Stack Trace: 
    SolutionState.CreateDiagnostic(MarkupOptions markupOptions, DiagnosticDescriptor defaultDiagnostic, ImmutableArray`1 supportedDiagnostics, ImmutableArray`1 fixableDiagnostics, String diagnosticId) line 432
    SolutionState.CreateDiagnosticForSpan(MarkupOptions markupOptions, DiagnosticDescriptor defaultDiagnostic, ImmutableArray`1 supportedDiagnostics, ImmutableArray`1 fixableDiagnostics, String diagnosticId, String filename, SourceText content, TextSpan span) line 381
    SolutionState.ProcessMarkupSources(IEnumerable`1 sources, IEnumerable`1 explicitDiagnostics, MarkupOptions markupOptions, DiagnosticDescriptor defaultDiagnostic, ImmutableArray`1 supportedDiagnostics, ImmutableArray`1 fixableDiagnostics, String defaultPath) line 337
    SolutionState.WithProcessedMarkup(MarkupOptions markupOptions, DiagnosticDescriptor defaultDiagnostic, ImmutableArray`1 supportedDiagnostics, ImmutableArray`1 fixableDiagnostics, String defaultPath) line 270
    <RunAsync>d__38.MoveNext() line 220
    --- End of stack trace from previous location where exception was thrown ---
    ...
```

For example:

```cs
        [Fact]
        public async Task DiagnosticAnalyzer_Basic()
        {
            await AnalyzeAsync(@"
using System;
using System.IO;
using System.Threading;
class C
{
    async void M()
    {
        byte[] buffer = { 0xBA, 0x5E, 0xBA, 0x11, 0xF0, 0x07, 0xBA, 0x11 };
        using (FileStream fs = new FileStream(""path.txt"", FileMode.Create))
        {
            await {|CA1834:fs.WriteAsync(buffer, 0, buffer.Length)|};
        }
    }
}
            ");
        }
```

Reply:
Yes, after looking more closely I think for this case we cannot use markup, but I'd like @sharwell to confirm.